### PR TITLE
CUDA: batch-invariant flash-attn for narrow widths on SM120

### PR DESCRIPTION
**Pull request by cuhawk**
### Problem

The flash-attention dispatch picks the narrowest template instance that fits `Q->ne[1]`. Speculative decoding varies `ne[1]` from step to step — 1 for the reference decode, 3–4 for the verification pass at the default `--spec-draft-n-max 3` — so one logical computation is served by two instances that differ in tile geometry, and therefore in KV loop and combine order. They cannot agree bitwise.

Speculative decoding is lossless under greedy sampling by construction, since the target model ratifies every drafted token. On SM120 today it is not: enabling it changes the generated text.

### Reproduction

`test_flash_attn_ext_batch_invariance` (added here) compares one batched pass against per-token serial passes with `memcmp` and `max_err() = 0`, over `n_kv {2048, 7168, 16384}` × `n_tokens {1, 2, 3, 4}`.

On an RTX 5080 Laptop, current master:

```
FLASH_ATTN_EXT_BATCH_INVARIANCE(n_tokens=3,n_kv=2048)    FAIL
FLASH_ATTN_EXT_BATCH_INVARIANCE(n_tokens=4,n_kv=2048)    FAIL
FLASH_ATTN_EXT_BATCH_INVARIANCE(n_tokens=3,n_kv=7168)    FAIL
FLASH_ATTN_EXT_BATCH_INVARIANCE(n_tokens=4,n_kv=7168)    FAIL
FLASH_ATTN_EXT_BATCH_INVARIANCE(n_tokens=3,n_kv=16384)   FAIL
FLASH_ATTN_EXT_BATCH_INVARIANCE(n_tokens=4,n_kv=16384)   FAIL
6/12 tests passed
```

`n_tokens` 1 and 2 pass; the failures are exactly the widths a 3-token draft produces.

End to end, Qwen3.8-27B-UD-Q3_K_XL, `--temp 0 --seed 1 -fa on`, 640 tokens, same prompt:

| run | vs. non-speculative reference |
|---|---|
| `--spec-type ngram-mod`, master | **diverges** — different text from the first paragraph on |
| `--spec-type ngram-mod`, this PR | **byte identical** |

A low-entropy prompt (verbatim code edit) produced identical output either way, so whether it is user-visible depends on how close the top-2 logits are. The determinism property is broken regardless.

### Fix

On Blackwell, route widths 3 and 4 to the 4-wide vector tile, which widths 1 and 2 already reduce to. That leaves a single instance across the whole 1..4 range. Columns past `ne[1]` are masked and `ntiles_x` stays 1 over that range, so launch geometry is unchanged. `LLAMA_SM120_FA_INVARIANT=0` restores the stock dispatch.

With the change: `12/12 tests passed`.

### Cost

`test-backend-ops perf`, same card, hs 256 / 24 q-heads / 4 kv-heads / kv 7168 / q8_0 KV. Medians of interleaved runs, first round discarded. Widths 1, 2, 5 and 8 are controls — the change must not touch them:

| nb | master | this PR | delta |
|---|---|---|---|
| 1 | 87.44 µs | 87.32 | −0.1% |
| 2 | 132.52 | 132.62 | +0.1% |
| **3** | 159.56 | 185.81 | **+16.5%** |
| **4** | 162.04 | 187.15 | **+15.5%** |
| 5 | 184.16 | 182.83 | −0.7% |
| 8 | 186.38 | 186.16 | −0.1% |

The cost is confined to the affected widths, and it is not anomalous: it brings nb=3/4 to ~186 µs, which is what nb=5 and nb=8 already cost. The 4-wide tile does 4-wide-tile work.

An alternative that pins one MMA instance across widths 1..4 instead was measured and rejected — it is free at widths 3–4 (+0.4%) but costs **+84.5% at nb=1** and +21.2% at nb=2, i.e. it taxes ordinary single-token decode.

### Testing

- `test-backend-ops test -b CUDA0`: **14607/14607 passed, 2/2 backends**, no failures.
- Compiles clean for `CMAKE_CUDA_ARCHITECTURES=75;86;120`, and runtime behaviour is identical to a 120-only build.
- Hardware: RTX 5080 Laptop (SM120) only. The change is gated to Blackwell and leaves every other architecture on the existing dispatch — I cannot test whether other architectures have the same property.

### CODE REVIEW

**Comment by JohannesGaessler** [comment]
According to the llama.cpp [AI usage policy](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md#ai-usage-policy):

>It is strictly prohibited to use AI to write your posts for you (bug reports, feature requests, pull request descriptions, Github discussions, responding to humans, ...).
